x/mongo/driver/topology/topology_options.go

```diff
@@ -7,451 +7,370 @@
-type Option func(*config) error
-
-type config struct {
-	mode                   MonitorMode
-	replicaSetName         string
-	seedList               []string
-	serverOpts             []ServerOption
-	cs                     connstring.ConnString // This must not be used for any logic in topology.Topology.
-	uri                    string
-	serverSelectionTimeout time.Duration
-	serverMonitor          *event.ServerMonitor
-	srvMaxHosts            int
-	srvServiceName         string
-	loadBalanced           bool
+// Config is used to construct a topology.
+type Config struct {
+	Mode                   MonitorMode
+	ReplicaSetName         string
+	SeedList               []string
+	ServerOpts             []ServerOption
+	ConnString             connstring.ConnString // This must not be used for any logic in topology.Topology.
+	URI                    string
+	ServerSelectionTimeout time.Duration
+	ServerMonitor          *event.ServerMonitor
+	SRVMaxHosts            int
+	SRVServiceName         string
+	LoadBalanced           bool
 }
 
-func newConfig(opts ...Option) (*config, error) {
-	cfg := &config{
-		seedList:               []string{"localhost:27017"},
-		serverSelectionTimeout: 30 * time.Second,
-	}
+func hasOptionalData(co options.ClientOptions) bool {
+	return co.Direct != nil && *co.Direct ||
+		co.ServerMonitor != nil ||
+		co.ReplicaSet != nil ||
+		co.ServerSelectionTimeout != nil ||
+		co.LoadBalanced != nil
+}
 
-	for _, opt := range opts {
-		if opt == nil {
-			continue
-		}
-		err := opt(cfg)
-		if err != nil {
-			return nil, err
-		}
+// convertToDriverAPIOptions converts a options.ServerAPIOptions instance to a driver.ServerAPIOptions.
+func convertToDriverAPIOptions(s *options.ServerAPIOptions) *driver.ServerAPIOptions {
+	driverOpts := driver.NewServerAPIOptions(string(s.ServerAPIVersion))
+	if s.Strict != nil {
+		driverOpts.SetStrict(*s.Strict)
 	}
-
-	return cfg, nil
+	if s.DeprecationErrors != nil {
+		driverOpts.SetDeprecationErrors(*s.DeprecationErrors)
+	}
+	return driverOpts
 }
 
-// WithConnString configures the topology using the connection string.
-func WithConnString(fn func(connstring.ConnString) connstring.ConnString) Option {
-	return func(c *config) error {
-		cs := fn(c.cs)
-		c.cs = cs
-
-		if cs.ServerSelectionTimeoutSet {
-			c.serverSelectionTimeout = cs.ServerSelectionTimeout
-		}
-
-		var connOpts []ConnectionOption
-
-		if cs.AppName != "" {
-			c.serverOpts = append(c.serverOpts, WithServerAppName(func(string) string { return cs.AppName }))
-		}
-
-		if cs.Connect == connstring.SingleConnect || (cs.DirectConnectionSet && cs.DirectConnection) {
-			c.mode = SingleMode
-		}
+type client interface {
+	GetClusterClock() *session.ClusterClock
+	SetServerAPI(*driver.ServerAPIOptions)
+}
 
-		c.seedList = cs.Hosts
+// newConfig will translate data from client options into a topology config for building non-default deployments.
+func newConfig(co *options.ClientOptions, client client) (*Config, error) {
```

Comment:
It may be simpler to move this `ClientOptions -> Config` logic to the `mongo` package. If you do that, you can get rid of the `client` interface because you have direct access to `GetClusterClock()` and `SetServerAPI`. Is there anything that would be made more complex or block moving this logic to the `mongo` package?

Reply:
Since the client package needs to build a deployment from the topology package and the topology tests require building configuration structs from client options, this would create a circular dependency between the `topology` package and the `mongo` package.

Divjot is suggesting we put the accessor data on the topology structures and set them on the client as needed: https://github.com/mongodb/mongo-go-driver/pull/1044#discussion_r945231313 . What are your thoughts on this approach?

Reply:
I like the approach @divjotarora suggests if moving to the `mongo` package will cause a circular dependency.

Reply:
Yeah, I see the source of the circular dependency now. Sounds like it's still important to be able to initialize a `Topology` from a connection string for some tests, which we'd have to stop doing if we moved this logic into the `mongo` package. Ideally any test that needs to construct a `Topology` would be able to do so without the `connstring -> Topology` logic, but that doesn't seem immediately feasible, so seems like a good idea to keep this logic in the `topology` package.

As far as the `topology.client` interface, I agree it should be removed. There are 2 functions in the `topology.client` interface which each handle a different piece of information.

## `GetClusterClock`
This can just be a parameter on `newConfig`. Callers that don't know the cluster clock can pass `nil`.

E.g.
```go
func NewConfig(co *options.ClientOptions, clock *session.ClusterClock) {
	// ...
}
```

## `SetServerAPI`
We can prevent the need to assign the `driver.ServerAPIOptions` back to the `Client` by exporting the function for converting an `options.ServerAPIOptions` -> `driver.ServerAPIOptions` (i.e. `convertToDriverAPIOptions`).

E.g.
```go
func (c *Client) configure(...) error {
	// ...
	if opts.ServerAPIOptions != nil {
		c.serverAPI = topology.ConvertToDriverAPIOptions(opts.ServerAPIOptions)
		// ...
}

 func NewConfig(co *options.ClientOptions, clock *session.ClusterClock) (*Config, error) {
	// ...
	if co.ServerAPIOptions != nil {
		serverAPI = ConvertToDriverAPIOptions(co.ServerAPIOptions)
		// ...
```

Reply:
Setting the same data in two separate places doesn't seem quite right to me. I think Divjot's solution is the most uniform. Alternatively we could add a function to parse the data from `topology.Config` as a one off:

```go
// x/mongo/driver/topology/server_options.go
func ServerAPIFromServerOptions(opts []ServerOption) *driver.ServerAPIOptions {
  return newServerConfig(opts...).serverAPI
}
```

```go
// mongo/client.go
func NewClient(opts ...*options.ClientOptions) (*Client, error) { 
  // ...
  client.serverAPI = topology.ServerAPIFromServerOptions(cfg.ServerOpts)
  // ...
}
```